apps/rush-lib/src/cli/RushPnpmCommandLine.ts

```diff
@@ -0,0 +1,259 @@
+import { PrintUtilities } from '@rushstack/terminal';
+
+import { RushConfiguration } from '../api/RushConfiguration';
+import { NodeJsCompatibility } from '../logic/NodeJsCompatibility';
+import { SpawnSyncReturns } from 'child_process';
+
+export interface ILaunchRushPnpmInternalOptions {
+  isManaged: boolean;
+  alreadyReportedNodeTooNewError?: boolean;
+}
+
+const RUSH_SKIP_CHECKS_PARAMETER: string = '--rush-skip-checks';
+
+enum CommandKind {
+  KnownSafe,
+  ShowWarning,
+  Blocked
+}
+
+export class RushPnpmCommandLine {
+  private static _knownCommandMap: Map<string, CommandKind> = new Map<string, CommandKind>([
+    ['add', CommandKind.ShowWarning],
+    ['audit', CommandKind.KnownSafe],
+    ['exec', CommandKind.KnownSafe],
+    ['import', CommandKind.Blocked],
+    ['install', CommandKind.Blocked],
+    [/* synonym */ 'i', CommandKind.Blocked],
+    ['install-test', CommandKind.Blocked],
+    [/* synonym */ 'it', CommandKind.Blocked],
+    ['link', CommandKind.ShowWarning],
+    [/* synonym */ 'ln', CommandKind.ShowWarning],
+    ['list', CommandKind.KnownSafe],
+    [/* synonym */ 'ls', CommandKind.KnownSafe],
+    ['outdated', CommandKind.KnownSafe],
+    ['pack', CommandKind.KnownSafe],
+    ['prune', CommandKind.KnownSafe],
+    ['publish', CommandKind.KnownSafe],
+    ['rebuild', CommandKind.KnownSafe],
+    [/* synonym */ 'rb', CommandKind.KnownSafe],
+    ['remove', CommandKind.ShowWarning],
+    [/* synonym */ 'rm', CommandKind.ShowWarning],
+    ['root', CommandKind.KnownSafe],
+    ['run', CommandKind.KnownSafe],
+    ['start', CommandKind.KnownSafe],
+    ['store', CommandKind.KnownSafe],
+    ['test', CommandKind.KnownSafe],
+    [/* synonym */ 't', CommandKind.KnownSafe],
+    ['unlink', CommandKind.ShowWarning],
+    ['update', CommandKind.ShowWarning],
+    [/* synonym */ 'up', CommandKind.ShowWarning]
+  ]);
+
+  public static launch(launcherVersion: string, options: ILaunchRushPnpmInternalOptions): void {
+    // Node.js can sometimes accidentally terminate with a zero exit code  (e.g. for an uncaught
+    // promise exception), so we start with the assumption that the exit code is 1
+    // and set it to 0 only on success.
+    process.exitCode = 1;
+
+    try {
+      // Are we in a Rush repo?
+      let rushConfiguration: RushConfiguration | undefined = undefined;
+      if (RushConfiguration.tryFindRushJsonLocation()) {
+        rushConfiguration = RushConfiguration.loadFromDefaultLocation({ showVerbose: true });
+      }
+
+      NodeJsCompatibility.warnAboutCompatibilityIssues({
+        isRushLib: true,
+        alreadyReportedNodeTooNewError: !!options.alreadyReportedNodeTooNewError,
+        rushConfiguration
+      });
+
+      if (!rushConfiguration) {
+        throw new Error(
+          'The "rush-pnpm" command must be executed in a folder that is under a Rush workspace folder'
+        );
+      }
+
+      if (rushConfiguration.packageManager !== 'pnpm') {
```

Comment:
Should this command be generalized? My original attempt at this made it a command (that belonged to Rush, not a top-level command like `rush-pnpm` is mind you) named `rush pm` so that it could be generalized out, much in the way that the `InstallManager` code is generalized but has package manager-specific implementations. I get that the goal of this new action is intended for PNPM, but would be nice if we didn't have to create `rush-yarn` or `rush-npm` in the future.

Reply:
🤔  Actually I feel like `rush-yarn` and `rush-npm` would actually be the right design. 

For example, if we made it generic like `rush-package-manager --do-thing install` then the CLI syntax would be polymorphic. You could never be sure whether the command is correct or not without first inspecting rush.json to determine which underlying tool will be parsing the CLI parameters.